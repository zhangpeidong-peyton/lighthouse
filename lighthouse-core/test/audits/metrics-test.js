/**
 * @license Copyright 2018 The Lighthouse Authors. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */
'use strict';

const MetricsAudit = require('../../audits/metrics.js');
const TTIComputed = require('../../computed/metrics/interactive.js');

const pwaTrace = require('../fixtures/traces/progressive-app-m60.json');
const pwaDevtoolsLog = require('../fixtures/traces/progressive-app-m60.devtools.log.json');

const lcpTrace = require('../fixtures/traces/lcp-m78.json');
const lcpDevtoolsLog = require('../fixtures/traces/lcp-m78.devtools.log.json');

const metricsAllFramesTrace = require('../fixtures/traces/frame-metrics-m89.json');
const metricsAllFramesDevtoolsLog = require('../fixtures/traces/frame-metrics-m89.devtools.log.json'); // eslint-disable-line max-len

const artifactsTrace = require('../results/artifacts/defaultPass.trace.json');
const artifactsDevtoolsLog = require('../results/artifacts/defaultPass.devtoolslog.json');

/* eslint-env jest */

describe('Performance: metrics', () => {
  it('evaluates valid input correctly', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input correctly (throttlingMethod=provided)', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'provided'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with lcp) correctly', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: lcpTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: lcpDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with lcp from all frames) correctly', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: metricsAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: metricsAllFramesDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'provided'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0]).toMatchSnapshot();
  });

  it('evaluates valid input (with CLS) correctly', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: artifactsTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: artifactsDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const {details} = await MetricsAudit.audit(artifacts, context);
    expect(details.items[0].cumulativeLayoutShift).toMatchInlineSnapshot(`0.42`);
    expect(details.items[0].observedCumulativeLayoutShift).toMatchInlineSnapshot(`0.42`);
  });

  it('evaluates valid input (with CLS from all frames) correctly', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: metricsAllFramesTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: metricsAllFramesDevtoolsLog,
      },
    };

    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const {details} = await MetricsAudit.audit(artifacts, context);
    expect(details.items[0].cumulativeLayoutShift).toBeCloseTo(0.0011);
    expect(details.items[0].observedCumulativeLayoutShift).toBeCloseTo(0.0011);
    expect(details.items[0].cumulativeLayoutShiftAllFrames).toBeCloseTo(0.459);
    expect(details.items[0].observedCumulativeLayoutShiftAllFrames).toBeCloseTo(0.459);
  });

  it('does not fail the entire audit when TTI errors', async () => {
    const artifacts = {
      traces: {
        [MetricsAudit.DEFAULT_PASS]: pwaTrace,
      },
      devtoolsLogs: {
        [MetricsAudit.DEFAULT_PASS]: pwaDevtoolsLog,
      },
    };

    const mockTTIFn = jest.spyOn(TTIComputed, 'request');
    mockTTIFn.mockRejectedValueOnce(new Error('TTI failed'));
    const context = {settings: {throttlingMethod: 'simulate'}, computedCache: new Map()};
    const result = await MetricsAudit.audit(artifacts, context);
    expect(result.details.items[0].interactive).toEqual(undefined);
  });
});
