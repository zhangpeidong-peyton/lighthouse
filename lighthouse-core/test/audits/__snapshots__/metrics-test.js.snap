// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Performance: metrics evaluates valid input (with lcp from all frames) correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0.0011656245471340055,
  "cumulativeLayoutShiftAllFrames": 0.4591700003057729,
  "estimatedInputLatency": 16,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 863,
  "firstCPUIdleTs": 23466886143,
  "firstContentfulPaint": 863,
  "firstContentfulPaintTs": 23466886143,
  "firstMeaningfulPaint": 863,
  "firstMeaningfulPaintTs": 23466886143,
  "interactive": 863,
  "interactiveTs": 23466886143,
  "largestContentfulPaint": 863,
  "largestContentfulPaintAllFrames": 683,
  "largestContentfulPaintAllFramesTs": 23466705983,
  "largestContentfulPaintTs": 23466886143,
  "maxPotentialFID": 16,
  "observedCumulativeLayoutShift": 0.0011656245471340055,
  "observedCumulativeLayoutShiftAllFrames": 0.4591700003057729,
  "observedDomContentLoaded": 596,
  "observedDomContentLoadedTs": 23466619325,
  "observedFirstContentfulPaint": 863,
  "observedFirstContentfulPaintTs": 23466886143,
  "observedFirstMeaningfulPaint": 863,
  "observedFirstMeaningfulPaintTs": 23466886143,
  "observedFirstPaint": 616,
  "observedFirstPaintTs": 23466639588,
  "observedFirstVisualChange": 609,
  "observedFirstVisualChangeTs": 23466632130,
  "observedLargestContentfulPaint": 863,
  "observedLargestContentfulPaintAllFrames": 683,
  "observedLargestContentfulPaintAllFramesTs": 23466705983,
  "observedLargestContentfulPaintTs": 23466886143,
  "observedLastVisualChange": 5881,
  "observedLastVisualChangeTs": 23471904130,
  "observedLoad": 673,
  "observedLoadTs": 23466696096,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 23466023130,
  "observedSpeedIndex": 1583,
  "observedSpeedIndexTs": 23467605703,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 23466023130,
  "observedTraceEnd": 6006,
  "observedTraceEndTs": 23472029453,
  "speedIndex": 1583,
  "speedIndexTs": 23467606130,
  "totalBlockingTime": 0,
}
`;

exports[`Performance: metrics evaluates valid input (with lcp) correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftAllFrames": 0,
  "estimatedInputLatency": 543,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 3790,
  "firstCPUIdleTs": undefined,
  "firstContentfulPaint": 2289,
  "firstContentfulPaintTs": undefined,
  "firstMeaningfulPaint": 2758,
  "firstMeaningfulPaintTs": undefined,
  "interactive": 4671,
  "interactiveTs": undefined,
  "largestContentfulPaint": 2758,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "maxPotentialFID": 1336,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftAllFrames": 0,
  "observedDomContentLoaded": 1513,
  "observedDomContentLoadedTs": 713038536140,
  "observedFirstContentfulPaint": 1122,
  "observedFirstContentfulPaintTs": 713038144775,
  "observedFirstMeaningfulPaint": 1122,
  "observedFirstMeaningfulPaintTs": 713038144775,
  "observedFirstPaint": 1122,
  "observedFirstPaintTs": 713038144775,
  "observedFirstVisualChange": 1105,
  "observedFirstVisualChangeTs": 713038128064,
  "observedLargestContentfulPaint": 1122,
  "observedLargestContentfulPaintAllFrames": undefined,
  "observedLargestContentfulPaintAllFramesTs": undefined,
  "observedLargestContentfulPaintTs": 713038144775,
  "observedLastVisualChange": 1722,
  "observedLastVisualChangeTs": 713038745064,
  "observedLoad": 2159,
  "observedLoadTs": 713039182071,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 713037023064,
  "observedSpeedIndex": 1393,
  "observedSpeedIndexTs": 713038416494,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 713037023064,
  "observedTraceEnd": 7416,
  "observedTraceEndTs": 713044439102,
  "speedIndex": 3681,
  "speedIndexTs": undefined,
  "totalBlockingTime": 1205,
}
`;

exports[`Performance: metrics evaluates valid input correctly (throttlingMethod=provided) 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftAllFrames": 0,
  "estimatedInputLatency": 17,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 1582,
  "firstCPUIdleTs": 225415754204,
  "firstContentfulPaint": 499,
  "firstContentfulPaintTs": 225414670885,
  "firstMeaningfulPaint": 783,
  "firstMeaningfulPaintTs": 225414955343,
  "interactive": 1582,
  "interactiveTs": 225415754204,
  "largestContentfulPaint": undefined,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "maxPotentialFID": 198,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftAllFrames": 0,
  "observedDomContentLoaded": 560,
  "observedDomContentLoadedTs": 225414732309,
  "observedFirstContentfulPaint": 499,
  "observedFirstContentfulPaintTs": 225414670885,
  "observedFirstMeaningfulPaint": 783,
  "observedFirstMeaningfulPaintTs": 225414955343,
  "observedFirstPaint": 499,
  "observedFirstPaintTs": 225414670868,
  "observedFirstVisualChange": 520,
  "observedFirstVisualChangeTs": 225414692015,
  "observedLargestContentfulPaint": undefined,
  "observedLargestContentfulPaintAllFrames": undefined,
  "observedLargestContentfulPaintAllFramesTs": undefined,
  "observedLargestContentfulPaintTs": undefined,
  "observedLastVisualChange": 818,
  "observedLastVisualChangeTs": 225414990015,
  "observedLoad": 2199,
  "observedLoadTs": 225416370913,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 225414172015,
  "observedSpeedIndex": 605,
  "observedSpeedIndexTs": 225414776724,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 225414172015,
  "observedTraceEnd": 12540,
  "observedTraceEndTs": 225426711887,
  "speedIndex": 605,
  "speedIndexTs": 225414777015,
  "totalBlockingTime": 48,
}
`;

exports[`Performance: metrics evaluates valid input correctly 1`] = `
Object {
  "cumulativeLayoutShift": 0,
  "cumulativeLayoutShiftAllFrames": 0,
  "estimatedInputLatency": 78,
  "estimatedInputLatencyTs": undefined,
  "firstCPUIdle": 4313,
  "firstCPUIdleTs": undefined,
  "firstContentfulPaint": 1337,
  "firstContentfulPaintTs": undefined,
  "firstMeaningfulPaint": 1553,
  "firstMeaningfulPaintTs": undefined,
  "interactive": 4348,
  "interactiveTs": undefined,
  "largestContentfulPaint": undefined,
  "largestContentfulPaintAllFrames": undefined,
  "largestContentfulPaintAllFramesTs": undefined,
  "largestContentfulPaintTs": undefined,
  "maxPotentialFID": 396,
  "observedCumulativeLayoutShift": 0,
  "observedCumulativeLayoutShiftAllFrames": 0,
  "observedDomContentLoaded": 560,
  "observedDomContentLoadedTs": 225414732309,
  "observedFirstContentfulPaint": 499,
  "observedFirstContentfulPaintTs": 225414670885,
  "observedFirstMeaningfulPaint": 783,
  "observedFirstMeaningfulPaintTs": 225414955343,
  "observedFirstPaint": 499,
  "observedFirstPaintTs": 225414670868,
  "observedFirstVisualChange": 520,
  "observedFirstVisualChangeTs": 225414692015,
  "observedLargestContentfulPaint": undefined,
  "observedLargestContentfulPaintAllFrames": undefined,
  "observedLargestContentfulPaintAllFramesTs": undefined,
  "observedLargestContentfulPaintTs": undefined,
  "observedLastVisualChange": 818,
  "observedLastVisualChangeTs": 225414990015,
  "observedLoad": 2199,
  "observedLoadTs": 225416370913,
  "observedNavigationStart": 0,
  "observedNavigationStartTs": 225414172015,
  "observedSpeedIndex": 605,
  "observedSpeedIndexTs": 225414776724,
  "observedTimeOrigin": 0,
  "observedTimeOriginTs": 225414172015,
  "observedTraceEnd": 12540,
  "observedTraceEndTs": 225426711887,
  "speedIndex": 1676,
  "speedIndexTs": undefined,
  "totalBlockingTime": 777,
}
`;
